### PR TITLE
TPI tutorial 2: with Docker

### PR DESCRIPTION
**Pull request by MKhalusova**
TODO devrel team:

nothing :see_no_evil: :point_down:

TODO @casperdcl:

- [x] update the publication date
- [x] improve [image](https://www.figma.com/file/3qDwiz0vLu6vK7XQzkqoug/DVC?node-id=307%3A4)?
- [x] add comments url: link to ~discord~ discuss
- [x] design own visuals (currently using an image from Unsplash)
- [x] depends on #3087
- [x] review & rewrite everything because of API & feature changes :see_no_evil:
  + [x] removed the AWS AMI bits - no longer needed
  + [x] fix bugs
  + [x] fix & update explanations
  + [x] remove "building custom docker containers" tutorial. It was long, distracting, difficult, and could be done in a part 3 follow-up instead. It also seems a bit pointless, there are a million other tutorials on this.
  + [x] update author because [`git fame -wMC`](https://github.com/casperdcl/git-fame#faqs) says I wrote ~80% of this, oops...
- [x] update https://github.com/iterative/blog-tpi-bees

### CODE REVIEW

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+title: Using Iterative Terraform Provider with Docker
+date: 2022-02-28
+description: >
+  In this tutorial you'll learn how you can move use an existing Docker image with 
+  Iterative Terraform Provider, as well as build your own Docker image.
+descriptionLong: >
+  In this tutorial you'll learn how you can move use an existing Docker image with 
+  Iterative Terraform Provider, as well as build your own Docker image.
+picture: 2022-02-28/unsplash-containers.jpg
+author: maria_khalusova
```
Comment (julieg18) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: The branch can be deployed successfully once we [add a author to `content/authors`](https://dvc.org/doc/user-guide/contributing/blog#adding-authors).

Comment (MKhalusova) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: Right, the author is added in a different PR which will be published before this one, so I'm not even sure if it makes sense to add here as well :D

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
+---
+title: Using Iterative Terraform Provider with Docker
+date: 2022-02-28
+description: >
+  In this tutorial you'll learn how you can move use an existing Docker image with 
+  Iterative Terraform Provider, as well as build your own Docker image.
+descriptionLong: >
+  In this tutorial you'll learn how you can move use an existing Docker image with 
+  Iterative Terraform Provider, as well as build your own Docker image.
```
Comment (jorgeorpinel) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:9]: Should the short desc be shorter or the long one longer? Or both

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 1 above)
Comment (jorgeorpinel) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: It makes some sense, so we can see the review app.

What's the other PR? Maybe that one is ready to merge

Comment (MKhalusova) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: Nope, it's not. It's the other tutorial that will be published after the release of TPI.
So here's the sequence. First, there will be a release blog post (currently in google docs), then the blog post with the tutorial #1, then this one. I'm building up a backlog of posts.

Comment (MKhalusova) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: Not sure why github decided to make that a link to an issue, please ignore that link.

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+in traditional software development, but also in machine learning. If you're not sure why it would be a good idea to use 
+Docker for machine learning development, check out 
+[this article from AWS](https://aws.amazon.com/blogs/opensource/why-use-docker-containers-for-machine-learning-development/) 
+that should answer this question for you. As a side note, using Docker to encapsulate your environment doesn't mean you won't
+ be using AMIs at all. In fact, AMI (Amazon Machine Image) is actually a required component of setting up an EC2, and at the 
+ minimum you will need an Ubuntu distribution AMI (which is a configurable default for the `iterative_task`). However, we will 
+ use Docker for the machine learning specific setup: drivers, GPU-specific libraries for accelerated math, and Python libraries. 
+ This separation will not only give you more control over the versions of the frameworks that you choose to use (or not use), 
+ but also over the cloud provider you provision the resources with. Your Docker image is not cloud provider specific while a 
+ Deep Learning AMI is a part of the AWS ecosystem.
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:38]:
```suggestion
 but also over the cloud provider you provision the resources with. Your Docker image is cloud provider agnostic.
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+ but also over the cloud provider you provision the resources with. Your Docker image is not cloud provider specific while a 
+ Deep Learning AMI is a part of the AWS ecosystem.  
+
+If you have a DevOps team, they can build a Docker image suited for your needs, so you'll only need to use it in your 
+`iterative_task` script. We'll show you how to do this in the first part of this tutorial.  But if you need (or want) to build 
+your own Docker image, we've got you too! The second part of this tutorial will illustrate some basics required for building 
+and using your own Docker image. Without further ado, let's get into it! 
+
+## Setting up environment for utilizing GPUs with Docker
+If you haven't read the [previous tutorial](2022-01-15-moving-local-experiments-to-the-cloud-with-iterative-terraform-provider.md)
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:46]: todo: update link

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+yet and are not familiar with Terraform, I would encourage you to check it out and learn some basics, i.e. how to let Terraform 
+know that you'll be using the Iterative Provider, how to initialize the working directory with `terraform init`, how to use 
+other essential Terraform commands (`plan`, `apply`, `refresh`, `show`, `destroy`). All these will apply here, we'll only modify
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:49]: did we mention `plan`?

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+If you haven't read the [previous tutorial](2022-01-15-moving-local-experiments-to-the-cloud-with-iterative-terraform-provider.md)
+yet and are not familiar with Terraform, I would encourage you to check it out and learn some basics, i.e. how to let Terraform 
+know that you'll be using the Iterative Provider, how to initialize the working directory with `terraform init`, how to use 
+other essential Terraform commands (`plan`, `apply`, `refresh`, `show`, `destroy`). All these will apply here, we'll only modify 
+the script part of the `main.tf`.
+
+Let's say we have a DevOps team that has carefully prepared a Docker image suitable for data science and machine learning. To 
+illustrate this, we can use a publicly available CML Docker image (`docker://iterativeai/cml`) provided by Iterative. This image 
+comes loaded with Python, CUDA, git, node and other essentials for full-stack data science. There are several versions of this 
+image, but for this example we'll take `docker://iterativeai/cml:0-dvc2-base1-gpu`, which includes Ubuntu 20.04, Python 3.8
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:55]:
```suggestion
comes loaded with Python, CUDA, git, node and other essentials for full-stack data science. There are
[several versions of this image](https://cml.dev/doc/self-hosted-runners#docker-images),
but for this example we'll take `docker://iterativeai/cml:0-dvc2-base1-gpu`, which includes Ubuntu 20.04, Python 3.8 
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+image, but for this example we'll take `docker://iterativeai/cml:0-dvc2-base1-gpu`, which includes Ubuntu 20.04, Python 3.8 
+(CUDA 11.0.3, CuDNN 8).
+
+Here's what the script is going to look like: 
+
+```hcl
+   script = <<-END
+   #!/bin/bash
+   sudo apt-get update
+   sudo apt-get install -y software-properties-common build-essential ubuntu-drivers-common
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:64]:
```suggestion
   sudo apt update -qq && sudo apt install -yqq software-properties-common build-essential ubuntu-drivers-common
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+   sudo ubuntu-drivers autoinstall
+   sudo curl -fsSL https://get.docker.com -o get-docker.sh && sudo sh get-docker.sh &&
+   sudo usermod -aG docker ubuntu
+   sudo setfacl --modify user:ubuntu:rw /var/run/docker.sock
+   curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
+   curl -s -L https://nvidia.github.io/nvidia-docker/ubuntu20.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+   sudo apt update && sudo apt install -y nvidia-docker2
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:71]:
```suggestion
   sudo apt update -qq && sudo apt install -yqq nvidia-docker2
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+Here's what the script is going to look like: 
+
+```hcl
+   script = <<-END
+   #!/bin/bash
+   sudo apt-get update
+   sudo apt-get install -y software-properties-common build-essential ubuntu-drivers-common
+   sudo ubuntu-drivers autoinstall
+   sudo curl -fsSL https://get.docker.com -o get-docker.sh && sudo sh get-docker.sh &&
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:66]:
```suggestion
   sudo curl -fsSL https://get.docker.com | sudo sh -
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+author: maria_khalusova
+#  todo: commentsUrl:
+tags:
+  - MLOps
+  - Iterative Terraform Provider
+  - AWS
+  - Docker
+---
+
+Recently we have [published a tutorial](2022-01-15-moving-local-experiments-to-the-cloud-with-iterative-terraform-provider.md)
```
Comment (jorgeorpinel) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:20]: Looks like a bad link probably?

Comment (jorgeorpinel) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:20]: p.s. another good reason to have a deployed review app, so the automated link checks can run

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+   sudo apt-get install -y software-properties-common build-essential ubuntu-drivers-common
+   sudo ubuntu-drivers autoinstall
+   sudo curl -fsSL https://get.docker.com -o get-docker.sh && sudo sh get-docker.sh &&
+   sudo usermod -aG docker ubuntu
+   sudo setfacl --modify user:ubuntu:rw /var/run/docker.sock
+   curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
+   curl -s -L https://nvidia.github.io/nvidia-docker/ubuntu20.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+   sudo apt update && sudo apt install -y nvidia-docker2
+   sudo systemctl restart docker
+   rm get-docker.sh
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:73]:
```suggestion
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+  
+   nvidia-smi
+   docker run --rm --gpus all -v "$PWD:/tpi" iterativeai/cml:0-dvc2-base1-gpu \
+       /bin/bash -c "cd /tpi; pip install -r requirements.txt; python src/train.py"
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:77]:
```suggestion
   docker run --rm --gpus all -v "$PWD:/tpi" -w /tpi iterativeai/cml:0-dvc2-base1-gpu \
       /bin/bash -c "pip install -r requirements.txt; python src/train.py"
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+   END
+```
+
+As you can see, we still need to make sure to install the required drivers first. Then in order to use a Docker image, we'll need 
+to install Docker itself, and sort out users, groups and permissions. Next, we'll install the `nvidia-docker` that will enable us 
+to run GPU-accelerated Docker containers. Yes, GPUs will make you work for them before you can enjoy the accelerated training!
+
+Finally, we can use Iterative's image to run our own container with this line: 
+```   docker run --rm --gpus all -v "$PWD:/tpi" iterativeai/cml:0-dvc2-base1-gpu \
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:86]: formatting issue?

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+       /bin/bash -c "cd /tpi; pip install -r requirements.txt; python src/train.py"
+```
+Let's unpack this line. 
+* `docker run`: this command will essentially create a container from the specified image, and then let us use it. 
+* `–rm flag` will make Docker automatically clean up the container and remove the file system when the container exits, 
+* `–gpus all` will allow us to use GPUs of the host machine from the container
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:92]:
```suggestion
* `--rm` flag will make Docker automatically clean up the container and remove the file system when the container exits, 
* `--gpus all` will allow us to use GPUs of the host machine from the container
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 1 above)
Comment (jorgeorpinel) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:11]: OK np. Please request my review here once https://github.com/iterative/dvc.org/pull/3087 is merged (and `master` merged back into this PR's branch 🙂). Thanks

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+FROM --platform=linux/amd64 nvidia/cuda:11.3.0-cudnn8-runtime-ubuntu20.04
+ 
+# Install wget and python
+RUN apt-get update && apt-get install -y wget python3-pip
+ 
+# Add the required packages to the environment
+COPY ./requirements.txt /tmp/requirements.txt
+RUN pip3 install -r /tmp/requirements.txt
+ 
+WORKDIR /app
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:128]: for consistency?

```suggestion
WORKDIR /tpi
```

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+```
+FROM --platform=linux/amd64 nvidia/cuda:11.3.0-cudnn8-runtime-ubuntu20.04
+ 
+# Install wget and python
+RUN apt-get update && apt-get install -y wget python3-pip
+ 
+# Add the required packages to the environment
+COPY ./requirements.txt /tmp/requirements.txt
+RUN pip3 install -r /tmp/requirements.txt
+
```
Comment (casperdcl) [inline content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:127]:
```suggestion
COPY requirements.txt .
RUN pip3 install -r requirements.txt && rm requirements.txt

```

**Review by casperdcl — CHANGES_REQUESTED** [review]
first round of feedback

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 10 above)
Comment (MKhalusova) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:20]: This is a link to the first tutorial that hasn't been published yet.

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 5 above)
Comment (MKhalusova) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:49]: Yes, I did mention `plan` in the first tutorial.

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 9 above)
Comment (DavidGOrtega) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md:66]: @casperdcl is this suggestion complete?

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md (hunk 2 above)
Comment (casperdcl) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md]: done

File: content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+To make it easy to get started, we used a pre-configured [Deep Learning AMI from AWS](https://aws.amazon.com/machine-learning/amis/) 
+to simplify the environment setup. In most cases, however, instead of using a pre-configured Deep Learning AMI, you will 
+ship your environment to a remote machine using [Docker](https://www.docker.com) which has become an essential tool not only 
+in traditional software development, but also in machine learning. If you're not sure why it would be a good idea to use 
+Docker for machine learning development, check out 
+[this article from AWS](https://aws.amazon.com/blogs/opensource/why-use-docker-containers-for-machine-learning-development/) 
+that should answer this question for you. As a side note, using Docker to encapsulate your environment doesn't mean you won't
+ be using AMIs at all. In fact, AMI (Amazon Machine Image) is actually a required component of setting up an EC2, and at the 
+ minimum you will need an Ubuntu distribution AMI (which is a configurable default for the `iterative_task`). However, we will
```
[the remark this answers is not in the document]
Comment (casperdcl) [reply content/blog/2022-02-28-using-iterative-terraform-provider-with-docker.md]: completely removed & rewritten

File: content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+[recommended in machine learning](https://aws.amazon.com/blogs/opensource/why-use-docker-containers-for-machine-learning-development/)
+as well as in traditional software development.
+
+[bees-part-1]: /blog/local-experiments-to-the-cloud-with-tpi
+[tpi]: https://github.com/iterative/terraform-provider-iterative
+
+<admon type="info">
+
+Using Docker to manage dependencies (e.g. Python packages) does not remove all
+other setup requirements. You'll still need docker itself installed, as well as
```
Comment (0x2b3bfa0) [inline content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md:45]:
```suggestion
other setup requirements. You'll still need Docker itself installed, as well as
```

`-Wdangerous-brand-eponym`

File: content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+  install the dependencies and run the training script.
+
+We can now call `terraform init` and `terraform apply` to provision
+infrastructure, sync data and code to it, set up the environment, and run the
+training process. If you'd like to tinker with this example you can
+[find it on GitHub](https://github.com/iterative/blog-tpi-bees/tree/docker).
+
+<admon type="tip">
+
+Don't forget to `terraform refresh && terraform show` to check the status, and
```
Comment (0x2b3bfa0) [inline content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md:118]: Worth mentioning a bit of the `TF_LOG_PROVIDER` awesomeness? 🦄🌈🤢

Comment (casperdcl) [reply content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md:118]: `I`DK

**Review by casperdcl — APPROVED** [review]

**Pull request merged**